extensions/indexes/lucene/src/org/exist/indexing/lucene/NodePathPattern.java

```diff
@@ -0,0 +1,230 @@
+        @Override
+        public boolean evaluate(NodePath2 nodePath, int elementIdx) {
+            return true;
+        }
+    };
+
+    static class SimpleAttrEqValuePredicate implements Predicate {
+        private final String attrName;
+        private final String attrVal;
+
+        SimpleAttrEqValuePredicate(String attrName, String attrVal) {
+            this.attrName = attrName;
+            this.attrVal  = attrVal;
+        }
+
+        @Override
+        public boolean evaluate(NodePath2 nodePath, int elementIdx) {
+            String val = nodePath.attribs(elementIdx).get(attrName);
+            return val != null ? val.equals(attrVal) : attrVal == null;
+        }
+    }
+
+
+    public NodePathPattern(Map<String, String> namespaces, String matchPattern) {
+        qnPath = new NodePath();
+        parseXPathExpression(namespaces, matchPattern);
+    }
+
+    public NodePathPattern(final QName qname) {
+        qnPath = new NodePath(qname);
+    }
+
+    private void parseXPathExpression(final Map<String, String> namespaces, final String matchPattern) {
+        //TODO : compute better length
+        final FastStringBuffer token = new FastStringBuffer(matchPattern.length());
+        int pos = 0;
+        while (pos < matchPattern.length()) {
+            final char ch = matchPattern.charAt(pos);
+            switch (ch) {
+                case '/':
+                    final String next = token.toString();
+                    token.setLength(0);
+                    if (next.length() > 0) {
+                        addSegment(namespaces, next);
+                    }
+                    if (matchPattern.charAt(++pos ) == '/') {
+                        qnPath.addComponent(NodePath.SKIP);
+                        predicates.add(CONST_TRUE_PREDICATE);
+                    }
+                    break;
+                default:
+                    token.append(ch);
+                    pos++;
+                    break;
+            }
+        }
+        if (token.length() > 0) {
+            addSegment(namespaces, token.toString());
+        }
+    }
+
+    private void  addSegment(final Map<String, String> namespaces, final String segment) {
+        String qname;
+        int predBeg = segment.indexOf('[');
+        Predicate pred = null;
+        if (predBeg >= 0) {
+            qname = segment.substring(0, predBeg);
+            pred = parsePredicate(segment.substring(predBeg));
+        }
+        else {
+            qname = segment.trim();
+            pred = CONST_TRUE_PREDICATE;
+        }
+
+        if ("*".equals(qname)) {
+            qnPath.addComponent(NodePath.WILDCARD);
+        }
+        else {
+            qnPath.addComponent(namespaces, qname);
+        }
+        predicates.add(pred);
+    }
+
+    private Predicate  parsePredicate(String input) {
+        if (!input.startsWith("[") || !input.endsWith("]") || input.charAt(1) != '@')
```

Comment:
what happens if string length == 0 ? or is that not possible?

Reply:
Not possible.
The only code snippet that invokes parsePredicate(...) is this:
    private void  addSegment(final Map<String, String> namespaces, final String segment) {
        String qname;
        int predBeg = segment.indexOf('[');
        Predicate pred = null;
        if (predBeg >= 0) {
            qname = segment.substring(0, predBeg);
            pred = parsePredicate(segment.substring(predBeg));
        }
//
But all in all -- perhaps, I should have added assert input.indexOf('[') >= 0
This way it would have been more informative